packages/next/src/server/lib/patch-fetch.ts

```diff
@@ -545,60 +550,69 @@ function createPatchedFetcher(
             }
             if (
               res.status === 200 &&
-              staticGenerationStore.incrementalCache &&
+              incrementalCache &&
               cacheKey &&
               isCacheableRevalidate
             ) {
-              const bodyBuffer = Buffer.from(await res.arrayBuffer())
-
-              try {
-                await staticGenerationStore.incrementalCache.set(
-                  cacheKey,
-                  {
-                    kind: 'FETCH',
-                    data: {
-                      headers: Object.fromEntries(res.headers.entries()),
-                      body: bodyBuffer.toString('base64'),
-                      status: res.status,
-                      url: res.url,
-                    },
-                    revalidate: normalizedRevalidate,
-                  },
-                  {
-                    fetchCache: true,
-                    revalidate,
-                    fetchUrl,
-                    fetchIdx,
-                    tags,
+              // Backport Notes: Previously this was gated behind `isCacheableRevalidate` because
+              // it was possible to flow into here in the HMR cache branch. Since that doesn't exist,
+              // we can simplify this to always set this to true if we're in here.
+              deferUnlockUntilCacheIsSet = true
+
+              res
+                .clone()
+                .arrayBuffer()
+                .then(async (arrayBuffer) => {
+                  const bodyBuffer = Buffer.from(arrayBuffer)
+
+                  const fetchedData = {
+                    headers: Object.fromEntries(res.headers.entries()),
+                    body: bodyBuffer.toString('base64'),
+                    status: res.status,
+                    url: res.url,
+                  }
+
+                  if (isCacheableRevalidate) {
+                    await incrementalCache.set(
+                      cacheKey!,
```

Comment:
Not 100% sure why this was erroring on the backport but not on canary unless I missed a guard somewhere. My guess is some TypeScript rules or version changed?

Reply:
Yes, this was fixed in [TypeScript 5.4](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-5-4.html#preserved-narrowing-in-closures-following-last-assignments).